pillarbox-demo/src/main/java/ch/srgssr/pillarbox/demo/ui/player/state/PictureInPictureButtonState.kt

```diff
@@ -0,0 +1,146 @@
+import androidx.core.util.Consumer
+
+/**
+ * Creates a [PictureInPictureButtonState] that is remembered across compositions.
+ *
+ * It is recommended to use [rememberPictureInPictureButtonState] with [PictureInPictureParams] for
+ * Android O and above for more control over the Picture-in-Picture behavior.
+ *
+ * @return A [PictureInPictureButtonState] instance.
+ */
+@Composable
+fun rememberPictureInPictureButtonState(): PictureInPictureButtonState {
+    val activity = LocalActivity.current as ComponentActivity
+    val pictureInPictureButtonState = remember(activity) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            Log.w("PictureInPictureButtonState", "Consider migrating to rememberPictureInPictureButtonState(PictureInPictureParams)")
+        }
+
+        PictureInPictureButtonStateBase(activity)
+    }
+
+    DisposableEffect(activity) {
+        pictureInPictureButtonState.startObserving()
+
+        onDispose {
+            pictureInPictureButtonState.stopObserving()
+        }
+    }
+
+    return pictureInPictureButtonState
+}
+
+/**
+ * Creates a [PictureInPictureButtonState] that is remembered across compositions.
+ *
+ * @param pictureInPictureParamsProvider A provider to get the parameters to use when entering Picture-in-Picture mode.
+ * @return A [PictureInPictureButtonState] instance.
+ */
+@Composable
+@RequiresApi(Build.VERSION_CODES.O)
+fun rememberPictureInPictureButtonState(pictureInPictureParamsProvider: () -> PictureInPictureParams): PictureInPictureButtonState {
+    val activity = LocalActivity.current as ComponentActivity
+    val pictureInPictureParamsProvider by rememberUpdatedState(pictureInPictureParamsProvider)
+    val pictureInPictureButtonState = remember(activity, pictureInPictureParamsProvider) {
+        PictureInPictureButtonStateApi26(activity, pictureInPictureParamsProvider)
+    }
+
+    DisposableEffect(activity, pictureInPictureParamsProvider) {
+        pictureInPictureButtonState.startObserving()
+
+        onDispose {
+            pictureInPictureButtonState.stopObserving()
+        }
+    }
+
+    return pictureInPictureButtonState
+}
+
+/**
+ * State that holds all interactions to correctly deal with a UI component representing a Picture-in-Picture button.
+ */
+interface PictureInPictureButtonState {
+    /**
+     * Whether Picture-in-Picture is available.
+     */
+    val isEnabled: Boolean
+
+    /**
+     * Whether the [Activity] is currently in Picture-in-Picture mode.
+     */
+    val isInPictureInPicture: Boolean
+
+    /**
+     * Enter Picture-in-Picture mode.
+     */
+    fun onClick()
+
+    /**
+     * Start observing the [Activity]'s [PictureInPictureModeChangedInfo] events.
+     */
+    fun startObserving()
+
+    /**
+     * Stop observing the [Activity]'s [PictureInPictureModeChangedInfo] events.
+     */
+    fun stopObserving()
+}
+
+private open class PictureInPictureButtonStateBase(
+    private val activity: ComponentActivity,
```

Comment:
I don't like much to store the Activity which can lead to memory leak. Can't we just put the activity as parameters of the methods `startObserving` and `stopObserving` ?

Other alternative are making `activity` nullable or using a weak reference `WeakReference<ComponentActivity>`.

Reply:
Good point! I'll look into it